text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+Modules in **Own Javascript** are allowed to use ECMA static `import` to resolve any **allowed dependency**, causing it to be included in the build whenever the importing module is included. This replaces `app.import`. Resolution follows prevailing Node rules. (This usually means the node_modules algorithm, but it could also mean Yarn PnP. The difference shouldn't matter if you are correctly declaring all your **allowed dependencies**.)
+
+Notice that a package’s **allowed dependencies** do not include the package itself. This is consistent with how Node resolution works. To import files from within your own package you must use relative paths. This is different from how run-time AMD module resolution has historically worked in Ember Apps. (`@embroider/compat` implements automatic adjustment for this case when compiling from v1 to v2).
+
+Modules in **Own Javascript** are allowed to use dynamic `import()`, and the specifiers have the same meanings as in static import. However, we impose a restriction on what syntax is supported inside `import()`. The only supported syntax inside `import()` is:
+
+- a string-literal
+- or a template string literal
+  - that begins with a static part
+  - where the static part is required to contain at least one `/`
+  - or at least two `/` when the path starts with `@`.
+
+This is designed to allow limited pattern matching of possible targets for your dynamic `import()`. The rules ensure that we can always tell which NPM package you are talking about (the `@` rule covers namespaced NPM packages, so you can't depend on `@ember/${whatever}` but you can depend on `@ember/translations/${lang}`). If your pattern matches zero files, we consider that a static build error. (For more background on the thought-process behind this feature, see [Embroider #198: Allow dynamic imports to accept more dynamic inputs](https://github.com/embroider-build/embroider/issues/198).
+
+Modules in **Own Javascript** are allowed to import template files. This is common in today’s addons (they import their own layout to set it on their Component class). The `.hbs` file extension resolves with lower priority than `.js` (meaning if you have both, the `.js` wins).
+
+Modules in **Own Javascript** are allowed to import JSON files. This is supported in Node and it's enabled by default in popular Javascript bundlers like Webpack, so it's common to encounter code on NPM that expects it to work. We interpret JSON as an ECMA module whose `default` export is the JSON object.
+
+### Own Javascript: Co-located Component Templates
+
+Modules in **Own Javascript** get no _automatic_ component template co-location. They should use Ember's `setComponentTemplate` explicitly. They can access their compiled templates either by importing them or by inlining them via the `hbs` macro provided by ember-cli-htmlbars.
+
+We're assuming that template co-location will be handled by the build step _before_ an addon is published to NPM in V2 format.
```

Comment:
I think the right place to address this is in the shared tooling. I've added to the RFC that we should ship an official package that gives addon authors tooling for this.

(Deciding which hook to use isn't in scope for the RFC because you can build your package using whatever bizarre system you want as long as the output conforms to the standrad, but shipping one good shared solution is definitely important!)

Reply:
Right, but I do want to make sure that there is _some_ hook that works with linking, installing from GitHub, doesn't have any undesirable side-effect (e.g. using postinstall and forcing a multi-minute wait for the build after running `yarn install`). At some point this was a mess and none of the option seem to work in all cases or have notable downsides. Not sure if it has changed since, but I certainly hope so. Maybe @rwjblue will know.